File: 6.Two Pointer/javascript_solution/7.js
function lifeboatsNeeded(weights, n, k) {
    // Step 1: Sort the array in ascending order
    weights.sort((a, b) => a - b);

    let left = 0; // Pointer to the lightest person
    let right = n - 1; // Pointer to the heaviest person
    let boats = 0;

    // Step 2: Use a two-pointer approach
    while (left <= right) {
        // Check if the lightest and heaviest person can share a boat
        if (weights[left] + weights[right] <= k) {
            left++; // Move to the next lightest person
            right--; // Move to the next heaviest person
        } else {
            right--; // Only take the heaviest person in a separate boat
        }
        
        // A boat is used in either case
        boats++;
    }

    return boats;
}

// Example usage:
let arr = [3, 5, 4, 3];
let n = arr.length;
let k = 5;
console.log(lifeboatsNeeded(arr, n, k)); // Output: 4











// Time Complexity Analysis
// Sorting the array → O(nlogn)
// The sort() function in JavaScript typically uses Timsort, which runs in O(n log n).
// Two-pointer traversal → O(n)
// We iterate through the array once, moving left and right pointers towards each other.
// Each element is processed at most once, making this step O(n).
// Overall Time Complexity:- O(nlogn)+O(n)=O(nlogn)
// Sorting dominates the complexity, so the function runs in O(n log n).



// Space Complexity Analysis
// Sorting is in-place → O(1)
// JavaScript’s sort() sorts the array in place, requiring no extra space.
// Two pointers and a counter → O(1)
// We use a few integer variables (left, right, boats), which take constant space.
// Overall Space Complexity:- O(1)
// The algorithm does not use extra space apart from input storage, making it space-efficient.


// Final Complexity Summary
// Time Complexity: O(n log n)
// Space Complexity: O(1)

File: 6.Two Pointer/python_solution/7.py
def lifeboats_needed(weights, n, k):
    # Step 1: Sort the list of weights in ascending order
    # This allows us to consider the lightest and heaviest person together.
    weights.sort()
    
    # Initialize two pointers:
    # 'left' starts from the beginning (lightest person)
    left = 0
    # 'right' starts from the end (heaviest person)
    right = n - 1
    
    # Counter to keep track of the number of boats needed
    boats = 0

    # Step 2: Use a while loop to pair people until all are assigned to boats
    while left <= right:
        # Check if the current lightest and heaviest person can share a boat
        if weights[left] + weights[right] <= k:
            # If they can share a boat, move both pointers:
            # Increase 'left' to point to the next lightest person
            left += 1
            # Decrease 'right' to point to the next heaviest person
            right -= 1
        else:
            # If they can't share a boat (i.e., their combined weight exceeds the limit),
            # only take the heavier person in the current boat (as the lighter person can't fit with them).
            # Move 'right' pointer to the next heaviest person
            right -= 1
        
        # In both cases (whether they share a boat or not), we used one boat
        boats += 1
    
    # Step 3: Return the total number of boats needed
    return boats



arr = [3, 5, 4,3] 
n = 4
k=5
res = lifeboats_needed(arr,n,k)
print(res)

